### PR TITLE
[floorplan] lane layout calculator

### PR DESCRIPTION
**Pull request by SharzyL**
- [floorplan] poc: lane layout calculator
- [floorplan] rust rewrite of lane layout calculator

### CODE REVIEW

**Comment by sequencer** [comment]
please squash commits and package?

File: lane-layout-gen/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+
+use rayon::prelude::*;
+
+#[derive(Parser, Debug)]
+struct Args {
+    #[arg(short)]
+    n: u64,
+
+    #[arg(short, long, default_value_t = 1)]
+    iter: isize,
```
Comment (sequencer) [inline lane-layout-gen/src/main.rs:18]: Make it optional and add an optional parameter: `search-to` to set a minimal target, if not get, search it forever.

File: lane-layout-gen/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+}
+
+fn main() {
+    let args = Args::parse();
+
+    let mut searcher = LayoutSearcher::new(args.n);
+    searcher.search(args.iter);
+
+    println!("best_d: {}", searcher.best_d);
```
Comment (sequencer) [inline lane-layout-gen/src/main.rs:27]: Use json for integration with Scala.

File: lane-layout-gen/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+    n: u64,
+
+    #[arg(short, long, default_value_t = 1)]
+    iter: isize,
+}
+
+fn main() {
+    let args = Args::parse();
+
+    let mut searcher = LayoutSearcher::new(args.n);
```
Comment (sequencer) [inline lane-layout-gen/src/main.rs:24]: consider adding shape constraints to floorplan in the next PRs.

**Review by sequencer — CHANGES_REQUESTED** [review]

**Review by sequencer — APPROVED** [review]

File: lane-layout-gen/src/main.rs [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+
+#[derive(Parser, Debug)]
+struct Args {
+    #[arg(short)]
+    n: u64,
+
+    #[arg(long)]
+    iter: isize,
+
+    #[arg(long, default_value_t = 1919810)]
```
Comment (sequencer) [inline lane-layout-gen/src/main.rs:20]:
```suggestion
    #[arg(long, default_value_t = 19260817)]
```

**Review by sequencer — CHANGES_REQUESTED** [review]

File: lane-layout-gen/default.nix [hunk 5]
```diff
@@ -0,0 +1,8 @@
+{ rustPlatform }:
+
+rustPlatform.buildRustPackage {
+  name = "lane-layout-gen";
+  src = ./.;
+  cargoHash = "sha256-9KJscqc//BMJ+H1e20n/kvPaSxrFuezKyXlOXeMajNk=";
```
Comment (sequencer) [inline lane-layout-gen/default.nix:6]:
```suggestion
  cargoHash = "sha256-XXvxSeNBpnYAWqmXxHIKIh4mHLm0+hYMsVho4dRgOi8=";
```

**Review by sequencer — CHANGES_REQUESTED** [review]

File: lane-layout-gen/src/main.rs [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+
+use rayon::prelude::*;
+
+#[derive(Parser, Debug)]
+struct Args {
+    #[arg(short)]
+    n: u64,
+
+    #[arg(long)]
+    iter: isize,
```
Comment (sequencer) [inline lane-layout-gen/src/main.rs:18]: should be optional as well. consider if `search_until_d` is set.

**Review by sequencer — CHANGES_REQUESTED** [review]

File: lane-layout-gen/src/main.rs (hunk 6 above)
Comment (sequencer) [reply lane-layout-gen/src/main.rs:18]: and `iter` doesn't honor `search_until_d`  
```
➜ ./result/bin/lane-layout-gen -n 2048 --iter 10000 --search-until-d 51
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 10000/10000 [03:23<00:00, 49.22it/s]find a layout with the max distance: 50
```

Comment (SharzyL) [reply lane-layout-gen/src/main.rs:18]: as expected. When search_until_d is set, the search does not stop until finding a small layout

**Pull request closed without merging**